storage/rocksdb/ha_rocksdb.h

```diff
@@ -1386,4 +1386,72 @@ struct Rdb_inplace_alter_ctx : public my_core::inplace_alter_handler_ctx {
   Rdb_inplace_alter_ctx(const Rdb_inplace_alter_ctx &);
   Rdb_inplace_alter_ctx &operator=(const Rdb_inplace_alter_ctx &);
 };
+
+/*
+  Helper class to control access/init to handlerton instance.
+  Contains a flag that is set if the handlerton is in an initialized, usable
+  state, plus a reader-write lock to protect it without serializing reads.
+  Since we don't have static initializers for the opaque mysql_rwlock type,
+  use constructor and destructor functions to create and destroy
+  the lock before and after main(), respectively.
+*/
+struct Rdb_hton_init_state {
+  struct Scoped_lock {
+    Scoped_lock(Rdb_hton_init_state& state, bool write) : m_state(state) {
+      if (write)
+        m_state.lock_write();
+      else
+        m_state.lock_read();
+    }
+    ~Scoped_lock() {
+      m_state.unlock();
+    }
+  private:
+    Rdb_hton_init_state& m_state;
+  };
+
+  Rdb_hton_init_state() : m_initialized(false) {
+    /*
+      m_rwlock can not be instrumented as it must be initialized before
+      mysql_mutex_register() call to protect some globals from race condition.
+    */
+    mysql_rwlock_init(0, &m_rwlock);
+  }
+
+  ~Rdb_hton_init_state() {
+    mysql_rwlock_destroy(&m_rwlock);
+  }
+
+  void lock_read() {
+    mysql_rwlock_rdlock(&m_rwlock);
+  }
+
+  void lock_write() {
+    mysql_rwlock_wrlock(&m_rwlock);
+  }
+
+  void unlock() {
+    mysql_rwlock_unlock(&m_rwlock);
+  }
+
+  /*
+    Must be called with either a read or write lock held, unable to enforce
+    behavior as mysql_rwlock has no means of determining if a thread has a lock
+  */
+  bool initialized() {
+    return m_initialized;
+  }
+
+  /*
+    Must be called with only a write lock held, unable to enforce behavior as
```

Comment:
Can be implemented with a thread_local flag set/cleared on a lock_write/unlock, but not sure if worth it

Reply:
True, and I considered that, but seemed to be well over the reinvent the wheel. I could have easily just wrapped the rwlock alone and kept the flag apart but I thought combining them, even in this 'broken' fashion made more sense from a readibility and scoping standpoint.